8248238: Implementation: JEP 388: Windows AArch64 Support

mo-beck opened the pull request:
This is a continuation of https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-August/009566.html
 
Changes since then:
* We've improved the write barrier as suggested by Andrew [1]
* The define-guards around R18 have been changed to `R18_RESERVED`. This will be enabled for Windows only for now but will be required for the upcoming macOS+Aarch64 [2] port as well.
* We've incorporated https://github.com/openjdk/jdk/pull/154 by @AntonKozlov in our PR for now and built the Windows-specific CPU feature detection on top of it.

[1] https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-August/009597.html
[2] https://openjdk.java.net/jeps/8251280

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8248238](https://bugs.openjdk.java.net/browse/JDK-8248238): Implementation: JEP 388: Windows AArch64 Support


### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Chris Plummer](https://openjdk.java.net/census#cjplummer) (@plummercj - **Reviewer**)
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**) ⚠️ Review applies to 68f61d60746b3a2785399810195af655d1b37a58
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**) ⚠️ Review applies to 398d764565f958a8999dd016c53c33cce78ebf19


### Contributors
 * Monica Beckwith `<mbeckwit@openjdk.org>`
 * Ludovic Henry `<luhenry@openjdk.org>`
 * Bernhard Urban-Forster `<burban@openjdk.org>`

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/212/head:pull/212`
`$ git checkout pull/212`

mo-beck commented:
/issue add 8248672,8248500,8248500,8248676,8248663,8248681,8248498,8248656,8248659,8248660,8248670,8248787,824794

mo-beck commented:
/contributor add @mo-beck
/contributor add @luhenry
/contributor add @lewurm

lewurm commented:
/issue add 8248672,8248500,8248500,8248676,8248663,8248681,8248498,8248656,8248659,8248660,8248670,8248787,8247941

mo-beck commented:
/issue add 8248672,8248500,8248500,8248676,8248663,8248681,8248498,8248656,8248659,8248660,8248670,8248787,8247941

mo-beck commented:
/issue remove 8247941

erikj79 reviewed (commented):
Build changes look good to me. I will take this branch for a spin.

erikj79 commented:
Our linux-aarch64 build fails with this:
cc: error: unrecognized command line option '-std=c++14'
when compiling build/linux-aarch64/buildjdk/hotspot/variant-server/libjvm/objs/precompiled/precompiled.hpp.gch

I'm trying to configure a windows-aarch64 build, but it fails on fixpath. Is this something you are also experiencing, and if so, how are you addressing it?

lewurm commented:
Hey @erikj79, thank you so much for giving it a try!

> Our linux-aarch64 build fails with this:
> cc: error: unrecognized command line option '-std=c++14'
> when compiling build/linux-aarch64/buildjdk/hotspot/variant-server/libjvm/objs/precompiled/precompiled.hpp.gch

Hmm, that's interesting. What environment is that exactly? What `configure` line are you using there? We have tested on such a system:
```console
$ cat /etc/issue
Ubuntu 19.10 \n \l
$ bash configure --with-boot-jdk=/home/beurba/work/jdk-16+13 --with-jtreg
$ make clean CONF=linux-aarch64-server-release
$ make images JOBS=255 LOG=info CONF=linux-aarch64-server-release
$ ./build/linux-aarch64-server-release/images/jdk/bin/java -XshowSettings:properties -version 2>&1 | grep aarch64
    java.home = /home/beurba/work/jdk/build/linux-aarch64-server-release/images/jdk
    os.arch = aarch64
    sun.boot.library.path = /home/beurba/work/jdk/build/linux-aarch64-server-release/images/jdk/lib
```
--------------------------------------------------------
> I'm trying to configure a windows-aarch64 build, but it fails on fixpath. Is this something you are also experiencing, and if so, how are you addressing it?

Yes. As far as I understand, the problem is that `fixpath.exe` isn't built properly when doing cross-compiling on Windows targets (as it hasn't been a thing so far). We use a workaround internally https://gist.github.com/lewurm/c099a4b5fcd8a182510cbdeebcb41f77 , but a proper solution is under discussion on build-dev: https://mail.openjdk.java.net/pipermail/build-dev/2020-July/027872.html

erikj79 commented on make/autoconf/toolchain.m4:
```diff
@@ -889,13 +893,22 @@ AC_DEFUN_ONCE([TOOLCHAIN_SETUP_BUILD_COMPILERS],
           BUILD_DEVKIT_TOOLCHAIN_PATH="$BUILD_DEVKIT_ROOT/bin"
         fi
-        PATH="$BUILD_DEVKIT_TOOLCHAIN_PATH:$BUILD_DEVKIT_EXTRA_PATH"
+        UTIL_PREPEND_TO_PATH([PATH],$BUILD_DEVKIT_TOOLCHAIN_PATH)
```
Here is a problem. In our linux cross compile build, we rely on the PATH only being completely overwritten with the paths from the devkit here. Otherwise the UTIL_REQUIRE_PROGS may find /usr/bin/cc before $BUILD_DEVKIT_TOOLCHAIN_PATH/gcc.

This is the reason my linux-aarch64 (cross compile) build failed. The system installed cc was too old to recognize the -stdc=c++14 argument.

erikj79 replied:
I assume you need the rest of the PATH on Windows.

lewurm replied:
> I assume you need the rest of the PATH on Windows.

Doesn't look like it actually. I've reverted it, thanks for catching it.

erikj79 replied:
Thanks, I tried the updated patch and it works now.

lewurm commented:
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> On 18/09/2020 11:14, Monica Beckwith wrote:
> 
> > This is a continuation of https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-August/009566.html
> 
> The diffs in assembler_aarch64.cpp are mostly spurious. Please try this.


Thank you Andrew. Is the goal to reduce the patch diff in `assembler_aarch64.cpp`? If so, we need to get rid of the retry in your patch to avoid additional calls to `random`, e.g. something like this (the diff for the generated part would look indeed nicer with that:  https://gist.github.com/lewurm/2e7b0e00447696c75e00febb83034ba1 ):

```patch
--- a/src/hotspot/cpu/aarch64/aarch64-asmtest.py
+++ b/src/hotspot/cpu/aarch64/aarch64-asmtest.py
@@ -13,6 +13,8 @@ class Register(Operand):

     def generate(self):
         self.number = random.randint(0, 30)
+        if self.number == 18:
+            self.number = 17
         return self

     def astr(self, prefix):
@@ -37,6 +39,8 @@ class GeneralRegisterOrZr(Register):

     def generate(self):
         self.number = random.randint(0, 31)
+        if self.number == 18:
+            self.number = 16
         return self

     def astr(self, prefix = ""):
@@ -54,6 +58,8 @@ class GeneralRegisterOrZr(Register):
 class GeneralRegisterOrSp(Register):
     def generate(self):
         self.number = random.randint(0, 31)
+        if self.number == 18:
+            self.number = 15
         return self

     def astr(self, prefix = ""):
@@ -1331,7 +1337,7 @@ generate(SpecialCases, [["ccmn",   "__ ccmn(zr, zr, 3u, Assembler::LE);",
                         ["st1w",   "__ sve_st1w(z0, __ S, p1, Address(r0, 7));",         "st1w\t{z0.s}, p1, [x0, #7, MUL VL]"],
                         ["st1b",   "__ sve_st1b(z0, __ B, p2, Address(sp, r1));",        "st1b\t{z0.b}, p2, [sp, x1]"],
                         ["st1h",   "__ sve_st1h(z0, __ H, p3, Address(sp, r8));",        "st1h\t{z0.h}, p3, [sp, x8, LSL #1]"],
-                        ["st1d",   "__ sve_st1d(z0, __ D, p4, Address(r0, r18));",       "st1d\t{z0.d}, p4, [x0, x18, LSL #3]"],
+                        ["st1d",   "__ sve_st1d(z0, __ D, p4, Address(r0, r17));",       "st1d\t{z0.d}, p4, [x0, x17, LSL #3]"],
                         ["ldr",    "__ sve_ldr(z0, Address(sp));",                       "ldr\tz0, [sp]"],
                         ["ldr",    "__ sve_ldr(z31, Address(sp, -256));",                "ldr\tz31, [sp, #-256, MUL VL]"],
                         ["str",    "__ sve_str(z8, Address(r8, 255));",                  "str\tz8, [x8, #255, MUL VL]"],
```

lewurm commented:
> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> On 21/09/2020 09:18, Bernhard Urban-Forster wrote:
> 
> > Thank you Andrew. Is the goal to reduce the patch diff in `assembler_aarch64.cpp`? If so, we need to get rid of the
> > retry in your patch to avoid additional calls to `random`, e.g. something like this (the diff for the generated part
> > would look indeed nicer with that:  https://gist.github.com/lewurm/2e7b0e00447696c75e00febb83034ba1 ):
> > [...]
> 
> Yes, better. Thanks.

Great, I've updated the PR. Thank you!

mo-beck commented:
Thanks, Andrew for catching that. I have made the needful changes.

-Monica

> _Mailing list message from [Andrew Haley](mailto:aph@redhat.com) on [build-dev](mailto:build-dev@openjdk.java.net):_
> 
> On 18/09/2020 11:14, Monica Beckwith wrote:
> 
> > This is a continuation of
> > https://mail.openjdk.java.net/pipermail/aarch64-port-dev/2020-August/009566.html
> > Changes since then:
> > * We've improved the write barrier as suggested by Andrew [1]
> 
> It's still wrong, I'm afraid. This is not a full barrier:
> 
> +#define FULL_MEM_BARRIER atomic_thread_fence(std::memory_order_acq_rel);
> 
> it is only StoreStore|LoadStore|LoadLoad, but you need StoreLoad as
> well. It might well be that you get the same DMB ISH instruction, but
> unless you use a StoreLoad barrier here it's theoretically possible
> for a compiler to reorder accesses so that a processor sees its own
> stores before other processors do. (And it's confusing for the reader
> too.)
> 
> Use:
> 
> +#define FULL_MEM_BARRIER atomic_thread_fence(std::memory_order_seq_cst);
> 
> See here:
> 
> https://en.cppreference.com/w/cpp/atomic/memory_order
> 
> memory_order_seq_cst "...plus a single total order exists in which all
> threads observe all modifications in the same order (see
> Sequentially-consistent ordering below)"
> 
> --
> Andrew Haley (he/him)
> Java Platform Lead Engineer
> Red Hat UK Ltd. <https://www.redhat.com>
> https://keybase.io/andrewhaley
> EAC8 43EB D3EF DB98 CC77 2FAD A5CD 6035 332F A671

dholmes-ora commented on src/hotspot/os/windows/os_windows.cpp:
```diff
@@ -2552,6 +2542,27 @@ LONG WINAPI topLevelExceptionFilter(struct _EXCEPTION_POINTERS* exceptionInfo) {
       }
 
+#ifdef _M_ARM64
+      // Unsafe memory access
```
I'm not at all clear why this unsafe memory access handling is for Aarch64 only?

lewurm replied:
Hum, this was already part of [JDK-8250810](https://github.com/openjdk/jdk/commit/a4eaf9536c272862b5ec856bf263679be09bddc9) / [JDK-8248817](https://github.com/openjdk/jdk/commit/257809d7440e87ac595d03b6c9a98d8f457f314c)  (see a couple lines below in this PR) without the `ifdef` for Aarch64, but I messed up rebasing on top of it. I removed it now, thanks for catching!

dholmes-ora commented on src/hotspot/share/runtime/stubRoutines.cpp:
```diff
@@ -394,7 +394,7 @@ void StubRoutines::initialize2() {
   // test safefetch routines
   // Not on Windows 32bit until 8074860 is fixed
-#if ! (defined(_WIN32) && defined(_M_IX86))
+#if ! (defined(_WIN32) && defined(_M_IX86)) && !defined(_M_ARM64)
```
The comment needs updating to refer to Aarch64.

lewurm replied:
This is actually not needed anymore, as it does work just fine on Windows+Aarch64. Presumably we have added it in the early days of the port when we haven't figured out exception handling quite yet.

Thanks for catching David.

dholmes-ora reviewed (approved):
I've mainly looked at the non-Aarch64 specific changes. A couple of minor comments below.

dholmes-ora commented:
@mo-beck This PR should not be associated with any JBS issues which are targeted at repo-aarch64-port. All such issues should have been closed by now when the associated code was pushed to aarch64-port repo. That was the whole point of creating separate issues so that they could be tracked in the porting repo. Now that all of that work should be complete the only issue that should remain open in relation to the Windows-Aarch64 port is JDK-8248238. Thanks.

plummercj reviewed (changes requested):
I looked at changes to existing SA files. These changes look fine.

I did not look at the new aarch64 SA files other than the copyright section. I assume they are clones of the x64 versions with some symbolic renaming. If there is any more than that and you'd like me to have a look, let me know. 

As for the copyright in the new SA files, I believe it is incorrect and needs to include Oracle. There are a number of other non-SA files that are new and also have the same copyright issue.

I also looked at src/jdk.attach/windows/classes/sun/tools/attach/AttachProviderImpl.java. It looks fine except it needs a copyright date update.

mo-beck commented:
@dholmes-ora, makes sense. I will do the needful. Thanks.

> @mo-beck This PR should not be associated with any JBS issues which are targeted at repo-aarch64-port. All such issues should have been closed by now when the associated code was pushed to aarch64-port repo. That was the whole point of creating separate issues so that they could be tracked in the porting repo. Now that all of that work should be complete the only issue that should remain open in relation to the Windows-Aarch64 port is JDK-8248238. Thanks.

lewurm commented:
@plummercj thank you for your feedback. I've updated the copyright in mentioned files.

VladimirKempik commented on src/hotspot/cpu/aarch64/register_aarch64.cpp:
```diff
@@ -41,7 +41,7 @@ const char* RegisterImpl::name() const {
     "rscratch1", "rscratch2",
     "r10", "r11", "r12", "r13", "r14", "r15", "r16",
-    "r17", "r18", "r19",
+    "r17", NOT_R18_RESERVED("r18") WIN64_ONLY("rtls"), "r19",
```
For me this line doesn't look good in case of expanding this functionality to macos-aarch64
as it's just the name of register and it's r18 on every platform except WIN64 and has nothing to do with reserving r18.

lewurm replied:
The idea is that the naming should suggest that `r18` shouldn't be used on that particular platform. Same is true for macOS, but the ABI docs suggest a different usage, hence we have something like that in our internal branch for macOS:
```suggestion
    "r17", NOT_R18_RESERVED("r18") WIN64_ONLY("rtls") MACOS_ONLY("rplatform"), "r19",
```
Are you suggesting it should rather be something like this eventually?
```suggestion
    "r17", LINUX_ONLY("r18") WIN64_ONLY("rtls") MACOS_ONLY("rplatform"), "r19",
```

VladimirKempik replied:
this looks better I think, if it's done right from beginning, we won't have to modify it later.
The Question is, can we do it ahead of JEP-391 ?
If we can't then maybe better to leave it this way for now:
WIN64_ONLY("rtls") NOT_WIN64("r18")

lewurm replied:
Let us go with the following in this PR as that's the extend of this PR:
```suggestion
    "r17", LINUX_ONLY("r18") WIN64_ONLY("rtls"), "r19",
```
We can update it accordingly in a PR for JEP-391.

VladimirKempik replied:
Ok, Great.

mo-beck commented:
/issue remove 8248672,8248500,8248500,8248676,8248663,8248681,8248498,8248656,8248659,8248660,8248670,8248787,8247941

mo-beck commented:
/integrate

magicus reviewed (approved):
Build changes look good.

lewurm commented:
@theRealAph okay, I've changed the string representation of `r18` to `"r18_tls"` on every platform.

plummercj commented:
> @plummercj thank you for your feedback. I've updated the copyright in mentioned files.

Changes look good. Thanks. I'm marking as "reviewed" for serviceability changes and copyrights.

mo-beck commented:
/integrate

mo-beck commented:
> The JEP is not yet targeted so we have to wait for that formality. But
> once that happens I can sponsor for you.
Thanks.
> 
> Also note that the PR references the wrong JEP so can you please edit
> the description to fix that.
Added JEP # (388) here and updated the JBS entry. 
After looking at JEPs 386, 377, and 379, I also did the following:
- listed JDK-8248238 as a sub-task for JDK-8248496
- added this PR link in a comment for the JEP.

As soon as the JEP is targetted, I will update the "Fix version" for the 'Implementation' (JDK-8248238) and ping you @dholmes-ora .
> 
> Meanwhile I'll see if I can take this for a spin through our internal
> testing.
Thanks so much.

Regards,
Monica

> 
> Cheers,
> David
> -----

dholmes-ora commented:
The initial comment still has this incorrect link:

[2] https://openjdk.java.net/jeps/8251280

Please edit the comment and fix the link.

dholmes-ora commented:
/sponsor

mo-beck commented:
> @mo-beck The initial comment still has this incorrect link:
> 
> [2] https://openjdk.java.net/jeps/8251280
> 
> Please edit the comment and fix the link.

That was a link to the macOS + Arm64 port. But I have removed it as it wasn't needed in the description of this implementation.
